launcher_utils/key_refresh_and_update_namespace_pubkey.go

```diff
@@ -0,0 +1,141 @@
+	// Validate the namespace format
+	for _, prefix := range prefixes {
+		if prefix == "" {
+			err := errors.New("Invalid empty prefix for public key update")
+			return nil, "", err
+		}
+		if prefix[0] != '/' {
+			err := errors.New("Prefix specified for public key update must start with a '/'")
+			return nil, "", err
+		}
+	}
+
+	// Generate the endpoint url that can update the public key of prefixes
+	fedInfo, err := config.GetFederation(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	registryEndpoint := fedInfo.RegistryEndpoint
+	if registryEndpoint == "" {
+		err = errors.New("No registry endpoint specified; try passing the `-f` flag specifying the federation name")
+		return nil, "", err
+	}
+
+	prefixPubKeyUpdateUrl, err := url.JoinPath(registryEndpoint, "api", "v1.0", "registry", "updateNamespacesPubKey")
+	if err != nil {
+		err = errors.Wrap(err, "Failed to construct public key update endpoint URL: %v")
+		return nil, "", err
+	}
+
+	// Obtain server's issuer private key
+	key, err := config.GetIssuerPrivateJWK()
+	if err != nil {
+		err = errors.Wrap(err, "Failed to obtain server's issuer private key")
+		return nil, "", err
+	}
+
+	return key, prefixPubKeyUpdateUrl, nil
+}
+
+func updateNamespacesPubKey(ctx context.Context, prefixes []string) error {
+	siteName := param.Xrootd_Sitename.GetString()
+
+	key, url, err := updateNamespacesPubKeyPrep(ctx, prefixes)
+	if err != nil {
+		return err
+	}
+	if err = registry.NamespacesPubKeyUpdate(key, prefixes, siteName, url); err != nil {
+		return err
+	}
+	return nil
+}
+
+func triggerNamespacesPubKeyUpdate(ctx context.Context) error {
+	extUrlStr := param.Server_ExternalWebUrl.GetString()
+	extUrl, _ := url.Parse(extUrlStr)
+	namespace := server_structs.GetOriginNs(extUrl.Host)
+	if err := updateNamespacesPubKey(ctx, []string{namespace}); err != nil {
+		log.Errorf("Error updating the public key of the registered origin namespace %s: %v", namespace, err)
+	}
+
+	originExports, err := server_utils.GetOriginExports()
+	if err != nil {
+		return err
+	}
+	originExportsNs := make([]string, len(originExports))
+	for i, export := range originExports {
+		originExportsNs[i] = export.FederationPrefix
+	}
+	if err := updateNamespacesPubKey(ctx, originExportsNs); err != nil {
+		log.Errorf("Error updating the public key of origin-exported namespace(s): %v", err)
+	}
+	return nil
+}
+
+// Check the directory containing .pem files regularly, load new private key(s)
+// For origin server, if new file(s) are detected, then register the new public key
+func LaunchIssuerKeysDirRefresh(ctx context.Context, egrp *errgroup.Group, modules server_structs.ServerType) {
+	egrp.Go(func() error {
+		ticker := time.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				log.Debugln("Stopping periodic check for private keys directory.")
+				return nil
+			case <-ticker.C:
```

Comment:
Switch this routine to using `server_utils.LaunchWatcherMaintenance` instead.  The `server_utils.LaunchWatcherMaintenance` does effectively the same thing *except* it also sets up a watcher on the directory and will get notified immediately about changes.